fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentSessionManager.java

```diff
@@ -61,23 +61,20 @@ public PersistentStorageSession getSession(final String sessionId) {
             throw new IllegalArgumentException("session id must be non-null");
         }
 
-        final PersistentStorageSession session = sessionMap.get(sessionId);
-
-        if (session != null) {
-            return session;
-        }
-
-        final PersistentStorageSession newSession = new OCFLPersistentStorageSession(sessionId,
+        return sessionMap.computeIfAbsent(sessionId, key -> new OCFLPersistentStorageSession(
+                key,
                 fedoraOcflIndex,
-                objectSessionFactory);
-        sessionMap.put(sessionId, newSession);
-        return newSession;
+                objectSessionFactory));
     }
 
     @Override
     public PersistentStorageSession getReadOnlySession() {
         if (this.readOnlySession == null) {
```

Comment:
Is this line needed as it is repeated inside the `synchronized (this) {` block? I feel like it was just an oversight?

Reply:
No synchronization is required.

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent(K,java.util.function.Function)

>  The entire method invocation is performed atomically, so the function is applied at most once per key. Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this map.

Reply:
Whoops, I got confused about which line you were referring to and thought you meant L64.

Reply:
Do you mean L72?

If so, then yes, both conditionals are needed. https://en.wikipedia.org/wiki/Double-checked_locking

Reply:
Ah okay, thanks for the explanation.